File: LearningAgent/src/Game/Game.java
package Game;

import java.util.ArrayList;
import java.util.List;

import Helper.Pair;
import Helper.Random;
import MKAgent.Agent;

public class Game {
	private Agent south;
	private Agent north;
	private Kalah kalah;
	private List<Move> moves;

	public Game(Board board, Agent agent1, Agent agent2) {
		this.kalah = new Kalah(board);

		moves = new ArrayList<Move>();
		if (Random.getRandomNumber(0, 2) == 0) {

			agent1.setSide(Side.SOUTH);
			this.south = agent1;
			agent2.setSide(Side.NORTH);
			this.north = agent2;
		} else {
			agent1.setSide(Side.NORTH);
			this.north = agent1;
			agent2.setSide(Side.SOUTH);
			this.south = agent2;

		}
	}

	public List<Move> getMoves() {
		return moves;
	}

	public Board getBoard() {
		return this.kalah.getBoard();
	}

	// method used to perform a full game and reports back list of moves performed
	// by
	// sides
	public Pair<Move[], Side> Play() {
		Move currentMove;

		// first the south make first move
		currentMove = south.play(kalah.getBoard(),true);
		// store move in array
		moves.add(currentMove);
		// apply move to kalah and update currentSide
		Side currentSide = kalah.makeMove(currentMove);
		
		
		// System.out.println(currentSide);
		if (!kalah.gameOver()) {

			// secound player make move
			currentMove = north.play(kalah.getBoard(),true);
			// store move in array
			moves.add(currentMove);

			// if move is swap:
			if (currentMove.getHole() < 0) {
				// swap players and update currentSide
				Agent temp = south;
				south = north;
				south.setSide(Side.SOUTH);
				north = temp;
				north.setSide(Side.NORTH);
			}
			else {
				// apply move to kalah and update currentSide
				currentSide = kalah.makeMove(currentMove);
			}
			// while the game not over
			while (!kalah.gameOver()) {
				// switch currentSide:
				switch (currentSide) {			
				// player 1 move:
				case SOUTH:
					// get player 1 move
					currentMove = south.play(kalah.getBoard(),false);
					//System.out.println(currentMove);
					// store move in array
					moves.add(currentMove);
					// apply move to kalah and update currentSide
					currentSide = kalah.makeMove(currentMove);
					// System.out.println(currentSide);

					// player 2 move:
					break;
				case NORTH:
					// get player 2 move
					currentMove = north.play(kalah.getBoard(),false);
					//System.out.println(currentMove);
					// store move in array
					moves.add(currentMove);
					// apply move to kalah and update currentSide
					currentSide = kalah.makeMove(currentMove);
					// System.out.println(currentSide);

				}
				 //System.out.println(currentMove);

			}
		}
		Side winner = getWinner(kalah.getBoard());
		//System.out.println(winner);
		return new Pair<Move[], Side>(moves.toArray(new Move[0]), winner);
	}

	private Side getWinner(Board board) {
		int totalSouth = board.getSeedsInStore(Side.SOUTH);
		int totalNorth = board.getSeedsInStore(Side.NORTH);

		if (totalSouth > totalNorth)
			return Side.SOUTH;
		else if (totalSouth < totalNorth)
			return Side.NORTH;
		else
			return null;

	}

}

File: Game/Game.py
import numpy as np

from Game.Side import *
from Game.Kalah import *
from Game.Move import *

class Game:
    def __init__(self, board, agent1, agent2):
        self.kalah = Kalah(board)
        
        self.moves = []
        if (np.random.randint(2) == 0):
            agent1.setSide(Side.SOUTH)
            self.south = agent1
            agent2.setSide(Side.NORTH)
            self.north = agent2
        else:
            agent1.setSide(Side.NORTH)
            self.north = agent1
            agent2.setSide(Side.SOUTH)
            self.south = agent2
        
    def getMoves(self):
        return self.moves
    
    def getBoard(self):
        return self.kalah.getBoard()
    
    @staticmethod
    def getWinnerStatic(board):
        totalSouth = board.getSeedsInStore(Side.SOUTH)
        totalNorth = board.getSeedsInStore(Side.NORTH)
        
        if (totalSouth > totalNorth):
            return Side.SOUTH
        elif (totalSouth < totalNorth):
            return Side.NORTH
        else:
            return None
    
    def play(self):
        currentMove = self.south.play(self.kalah.getBoard(), True)
        self.moves.append(currentMove)
        currentSide = self.kalah.makeMove(currentMove);
        
        if (not self.kalah.gameOver()):
            currentMove = self.north.play(self.kalah.getBoard(), True)
            self.moves.append(currentMove)
            
            if (currentMove.getHole() < 0):
                temp = self.south
                self.south = self.north
                self.south.setSide(Side.SOUTH)
                self.north = temp
                self.north.setSide(Side.NORTH)
            else:
                currentSide = self.kalah.makeMove(currentMove)
                
            while(not self.kalah.gameOver()):
                if currentSide == Side.SOUTH:
                    currentMove = self.south.play(self.kalah.getBoard(), False)
                    self.moves.append(currentMove)
                    currentSide = self.kalah.makeMove(currentMove)
                else:
                    currentMove = self.north.play(self.kalah.getBoard(), False)
                    self.moves.append(currentMove)
                    currentSide = self.kalah.makeMove(currentMove)
        
        winner = Game.getWinnerStatic(self.kalah.getBoard());
        return (self.moves, winner)